tee_supplicant/src/tee_supplicant.c

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/tee.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <optee_msg_supplicant.h>
+#include <tee_client_api.h>
+#include <teec_ta_load.h>
+
+LOG_MODULE_REGISTER(tee_supplicant);
+
+#define TEE_SUPP_THREAD_PRIO	7
+
+#define TEE_REQ_PARAM_MAX	5
+
+static struct k_thread main_thread;
+static K_THREAD_STACK_DEFINE(main_stack, 8192);
+
+struct suppl_shm {
+	int id;
+	void *p;
+	size_t size;
+	bool registered;
+	struct suppl_shm *next;
+};
+
+static K_MUTEX_DEFINE(shm_mutex);
+static struct suppl_shm *shm_head;
+
+static void tee_supp_mutex_lock(struct k_mutex *mu);
+static void tee_supp_mutex_unlock(struct k_mutex *mu);
+
+#define MEMREF_SHM_ID(p)	((p)->c)
+#define MEMREF_SHM_OFFS(p)	((p)->a)
+#define MEMREF_SIZE(p)		((p)->b)
+
+struct tee_supp_request {
+	uint32_t cmd;
+	uint32_t num_param;
+	struct tee_param params[TEE_REQ_PARAM_MAX];
+};
+
+struct tee_supp_response {
+	uint32_t ret;
+	uint32_t num_param;
+	struct tee_param params[TEE_REQ_PARAM_MAX];
+};
+
+union tee_supp_msg {
+	struct tee_supp_request req;
+	struct tee_supp_response rsp;
+};
+
+struct param_value {
+	uint64_t a;
+	uint64_t b;
+	uint64_t c;
+};
+
+static int receive_request(const struct device *dev, struct tee_supp_request *ts_req)
+{
+	int rc = tee_suppl_recv(dev, &ts_req->cmd, &ts_req->num_param, ts_req->params);
+
+	if (rc) {
+		LOG_ERR("TEE supplicant receive failed, rc = %d", rc);
+	}
+
+	return rc;
+}
+
+static int send_response(const struct device *dev, struct tee_supp_response *rsp)
+{
+	int rc = tee_suppl_send(dev, rsp->ret, rsp->num_param, rsp->params);
+
+	if (rc) {
+		LOG_ERR("TEE supplicant send response failed, rc = %d", rc);
+	}
+
+	return rc;
+}
+
+static int get_value(size_t num_params, struct tee_param *params,
+		     const uint32_t idx, struct param_value **value)
+{
+	if (idx >= num_params) {
+		return -1;
+	}
+
+	switch (params[idx].attr & TEE_PARAM_ATTR_TYPE_MASK) {
+	case TEE_PARAM_ATTR_TYPE_VALUE_INPUT:
+	case TEE_PARAM_ATTR_TYPE_VALUE_OUTPUT:
+	case TEE_PARAM_ATTR_TYPE_VALUE_INOUT:
+		*value = (void *)&params[idx].a;
+		return 0;
+	default:
+		return -1;
+	}
+}
+
+static struct suppl_shm *find_tshm(int id)
+{
+	struct suppl_shm *tshm = NULL;
+
+	tee_supp_mutex_lock(&shm_mutex);
+
+	tshm = shm_head;
+	while (tshm && tshm->id != id) {
+		tshm = tshm->next;
+	}
+
+	tee_supp_mutex_unlock(&shm_mutex);
+
+	return tshm;
+}
+
+/* Get parameter allocated by secure world */
+static int get_param(size_t num_params, struct tee_param *params,
```

Comment:
It will be nice to rename and/or document this function.

It has quite complex logic and it is not obvious what exactly it does

Reply:
This function removed